src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,190 @@
+        if (index == size) {
+            linkLast(newNode);
+        } else if (index == 0) {
+            linkFirst(newNode);
+        } else {
+            link(newNode, index);
+        }
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T item : list) {
+            add(item);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index);
+        return getNodeByIndex(index).value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndex(index);
+        Node<T> newNode = getNodeByIndex(index);
+        T oldValue = newNode.value;
+        newNode.value = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index);
+        return unlink(getNodeByIndex(index));
     }
 
     @Override
     public boolean remove(T object) {
-        return false;
+        int removeByIndex = getIndexIfContains(object);
+        if (removeByIndex == -1) {
+            return false;
+        }
+        unlink(getNodeByIndex(removeByIndex));
+        return true;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void linkLast(Node<T> newNode) {
+        if (head == null) {
+            head = tail = newNode;
+        } else {
+            tail.next = newNode;
+            newNode.prev = tail;
+            tail = newNode;
+        }
+    }
+
+    private void linkFirst(Node<T> newNode) {
+        newNode.next = head;
+        head.prev = newNode;
+        head = newNode;
+    }
+
+    private void link(Node<T> newNode, int index) {
+        Node<T> previous = getNodeByIndex(index - 1);
+        newNode.next = previous.next;
+        previous.next.prev = newNode;
+        previous.next = newNode;
+        newNode.prev = previous;
+    }
+
+    private int getIndexIfContains(T object) {
+        int index = -1;
+        Node<T> current = head;
+        for (int i = 0; i < size; i++) {
+            if (current.value == object || current.value != null && current.value.equals(object)) {
+                return i;
+            }
+            current = current.next;
+        }
+        return index;
+    }
+
+    private T unlink(Node<T> deletedNode) {
+        if (size == 1) {
+            head = tail = null;
+            size--;
+            return deletedNode.value;
+        }
+        if (deletedNode.prev == null) {
+            head = head.next;
+            head.prev = null;
+            size--;
+            return deletedNode.value;
+        }
+        if (deletedNode.next == null) {
+            tail.prev.next = null;
+            tail = tail.prev;
+            size--;
+            return deletedNode.value;
+        }
```

Comment:
size--; can be placed here and removed from if statements and unlinkWithin method